components/server/ee/src/prebuilds/incremental-prebuilds-service.ts

```diff
@@ -0,0 +1,175 @@
+/**
```

Comment:
@jankeromnes Any changes in this file? Or is it a "pure move"? :thinking:

Reply:
This is a new file, composed of 3 pre-existing code sections. Here are the relevant diffs:

<details><summary>getCommitHistoryForContext diff</summary>

```diff
--- get-commit-history-1        2022-10-18 09:42:59.685989248 +0000
+++ get-commit-history-2        2022-10-18 09:41:19.713994670 +0000
@@ -1,18 +1,18 @@
 const maxDepth = this.config.incrementalPrebuilds.commitHistory;
 const hostContext = this.hostContextProvider.get(context.repository.host);
 const repoProvider = hostContext?.services?.repositoryProvider;
-if (repoProvider) {
-    prebuildContext.commitHistory = await repoProvider.getCommitHistory(
+if (!repoProvider) {
+    return {};
+}
+const history: WithCommitHistory = {};
+history.commitHistory = await repoProvider.getCommitHistory(
         user,
         context.repository.owner,
         context.repository.name,
         context.revision,
         maxDepth,
     );
-    if (
-        context.additionalRepositoryCheckoutInfo &&
-        context.additionalRepositoryCheckoutInfo.length > 0
-    ) {
+if (context.additionalRepositoryCheckoutInfo && context.additionalRepositoryCheckoutInfo.length > 0) {
         const histories = context.additionalRepositoryCheckoutInfo.map(async (info) => {
             const commitHistory = await repoProvider.getCommitHistory(
                 user,
@@ -26,5 +26,6 @@ if (repoProvider) {
                 commitHistory,
             };
         });
-        prebuildContext.additionalRepositoryCommitHistories = await Promise.all(histories);
+    history.additionalRepositoryCommitHistories = await Promise.all(histories);
     }
```

</details>

<details><summary>findGoodBaseForIncrementalBuild diff</summary>

```diff
--- find-good-base-1    2022-10-18 09:51:32.557961085 +0000
+++ find-good-base-2    2022-10-18 09:41:40.157993561 +0000
@@ -1,23 +1,23 @@
-if (context.commitHistory && context.commitHistory.length > 0) {
+if (!history.commitHistory || history.commitHistory.length < 1) {
+    return;
+}
+
+const { config } = await this.configProvider.fetchConfig({}, user, context);
+const imageSource = await this.imageSourceProvider.getImageSource({}, user, context, config);
+
     // Note: This query returns only not-garbage-collected prebuilds in order to reduce cardinality
     // (e.g., at the time of writing, the Gitpod repository has 16K+ prebuilds, but only ~300 not-garbage-collected)
-    const recentPrebuilds = await this.db.trace({ span }).findPrebuildsWithWorkpace(commitContext.repository.cloneUrl);
-    const loggedContext = filterForLogging(context);
+const recentPrebuilds = await this.workspaceDB.findPrebuildsWithWorkpace(context.repository.cloneUrl);
     for (const recentPrebuild of recentPrebuilds) {
         if (
-            !(await this.isGoodBaseforIncrementalPrebuild(
-                context,
+        await this.isGoodBaseforIncrementalBuild(
+            history,
                 config,
                 imageSource,
                 recentPrebuild.prebuild,
                 recentPrebuild.workspace,
-            ))
+        )
         ) {
-            log.debug({ userId: user.id }, "Not using incremental prebuild base", {
-                candidatePrebuildId: recentPrebuild.prebuild.id,
-                context: loggedContext,
-            });
-            continue;
-        }
+        return recentPrebuild.prebuild;
     }
 }
```

</details>

<details><summary>isGoodBaseForIncrementalBuild diff</summary>

```diff
--- is-good-base-1      2022-10-18 09:44:50.477983240 +0000
+++ is-good-base-2      2022-10-18 09:42:00.273992470 +0000
@@ -1,7 +1,7 @@
-if (!context.commitHistory || context.commitHistory.length === 0) {
+if (!history.commitHistory || history.commitHistory.length === 0) {
     return false;
 }
-if (!CommitContext.is(candidate.context)) {
+if (!CommitContext.is(candidateWorkspace.context)) {
     return false;
 }
 
@@ -11,23 +11,26 @@ if (candidatePrebuild.state !== "availab
 }
 
 // we are only considering full prebuilds
-if (!!candidate.basedOnPrebuildId) {
+if (!!candidateWorkspace.basedOnPrebuildId) {
     return false;
 }
 
-const candidateCtx = candidate.context;
-if (candidateCtx.additionalRepositoryCheckoutInfo?.length !== context.additionalRepositoryCommitHistories?.length) {
+if (
+    candidateWorkspace.context.additionalRepositoryCheckoutInfo?.length !==
+    history.additionalRepositoryCommitHistories?.length
+) {
     // different number of repos
     return false;
 }
 
-if (!context.commitHistory.some((sha) => sha === candidateCtx.revision)) {
+const candidateCtx = candidateWorkspace.context;
+if (!history.commitHistory.some((sha) => sha === candidateCtx.revision)) {
     return false;
 }
 
 // check the commits are included in the commit history
-for (const subRepo of candidateCtx.additionalRepositoryCheckoutInfo || []) {
-    const matchIngRepo = context.additionalRepositoryCommitHistories?.find(
+for (const subRepo of candidateWorkspace.context.additionalRepositoryCheckoutInfo || []) {
+    const matchIngRepo = history.additionalRepositoryCommitHistories?.find(
         (repo) => repo.cloneUrl === subRepo.repository.cloneUrl,
     );
     if (!matchIngRepo || !matchIngRepo.commitHistory.some((sha) => sha === subRepo.revision)) {
@@ -36,10 +39,10 @@ for (const subRepo of candidateCtx.addit
 }
 
 // ensure the image source hasn't changed (skips older images)
-if (JSON.stringify(imageSource) !== JSON.stringify(candidate.imageSource)) {
+if (JSON.stringify(imageSource) !== JSON.stringify(candidateWorkspace.imageSource)) {
     log.debug(`Skipping parent prebuild: Outdated image`, {
         imageSource,
-        parentImageSource: candidate.imageSource,
+        parentImageSource: candidateWorkspace.imageSource,
     });
     return false;
 }
@@ -55,7 +58,7 @@ const filterPrebuildTasks = (tasks: Task
         )
         .filter((task) => Object.keys(task).length > 0);
 const prebuildTasks = filterPrebuildTasks(config.tasks);
-const parentPrebuildTasks = filterPrebuildTasks(candidate.config.tasks);
+const parentPrebuildTasks = filterPrebuildTasks(candidateWorkspace.config.tasks);
 if (JSON.stringify(prebuildTasks) !== JSON.stringify(parentPrebuildTasks)) {
     log.debug(`Skipping parent prebuild: Outdated prebuild tasks`, {
         prebuildTasks,
```

</details>

Reply:
Would have been great to see this diff surfaced in the PR, ideally by making the move a separate PR, or at least a separate commit. :confused: 
Only now I notice that `fetchConfig` was pulled into `findGoodBaseForIncrementalBuild`, which feels a bit strange, because now there is a disconnect between these two lines [[1](https://github.com/gitpod-io/gitpod/blob/5687600aab763adc2562b31163dd337c9332aa4e/components/server/ee/src/workspace/workspace-factory.ts#L118), [2](https://github.com/gitpod-io/gitpod/blob/5687600aab763adc2562b31163dd337c9332aa4e/components/server/ee/src/workspace/workspace-factory.ts#L151-L158)]. :thinking: 

Something for a follow-up?

Reply:
Thanks @geropl!

> Would have been great to see this diff surfaced in the PR, ideally by making the move a separate PR, or at least a separate commit. 😕

True, I agree that code moves and code changes would ideally be made in separate commits. 💯 (But unfortunately, I squashed my commits, to make resolving some merge conflicts easier.)

> Only now I notice that `fetchConfig` was pulled into `findGoodBaseForIncrementalBuild`, which feels a bit strange, because now there is a disconnect between these two lines [[1](https://github.com/gitpod-io/gitpod/blob/5687600aab763adc2562b31163dd337c9332aa4e/components/server/ee/src/workspace/workspace-factory.ts#L118), [2](https://github.com/gitpod-io/gitpod/blob/5687600aab763adc2562b31163dd337c9332aa4e/components/server/ee/src/workspace/workspace-factory.ts#L151-L158)]. :thinking: 

Indeed, you're right about the `fetchConfig` being moved. And this has the unfortunate consequence of calling `fetchConfig` twice now ([[1](https://github.com/gitpod-io/gitpod/blob/9dffa5d337eb4ed1787084d1aa90f6cee2a7560d/components/server/ee/src/workspace/workspace-factory.ts#L118)], [[2](https://github.com/gitpod-io/gitpod/blob/9dffa5d337eb4ed1787084d1aa90f6cee2a7560d/components/server/ee/src/prebuilds/incremental-prebuilds-service.ts#L76)]). This could be resolved by allowing to pass an optional `config` to `findGoodBaseForIncrementalBuild`.

However, I don't understand what you mean by "disconnect". These two lines still do what was intended, right? (I.e., we fetch the `config` for the current context, then we find a potentially older prebuild, but we "fix" the created workspace by forcing its `config` back to the most recent one.)

FYI, this was done to fix a bug where, during incremental prebuilds, a prebuild somehow got a `config` with an outdated `command`, and the new workspace would run the outdated `command` on start-up even though the `.gitpod.yml` in the workspace and Git history both had the more recent `command`.

I think that bug is still adequately handled in the new version of the code. Does this explanation resolve the "disconnect" you saw?

Reply:
With "disconnect" I mean that we should be using the exact same `config` being used for the prebuild and for identifying the prebuild.
Now that we're doing it twice that might for some CommitContexts (e.g., PR/branch context) lead to situations where config is out of sync here. This a) makes it harder to reason about the code, and b) opens up possibility for current (or future) bugs.

Reply:
Thanks for explaining! I don't fully understand how the configs could get out-of-sync (because we only fetch a read-only config, i.e. a `.gitpod.yml` from the repo, for the single context that we currently want to open, regardless of which prebuild or incremental prebuild we end up using -- we do occasionally fetch the same config multiple times, but we never modify it).

Maybe, returning to fetching it just once here, and passing it along to `findGoodBaseForIncrementalBuild` would solve your concern? (I.e. then we keep exactly one `config` object in this flow as before.)